eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs

```diff
@@ -3,63 +3,110 @@
+import Cardano.Ledger.Core (Era (EraCrypto), EraPParams, EraRule, PParams)
+import Cardano.Ledger.Credential (Credential)
+import Cardano.Ledger.Keys (KeyHash, KeyRole (CommitteeColdKey, Voting))
 import Control.DeepSeq (NFData)
-import Control.State.Transition (
+import Control.State.Transition.Extended (
   BaseM,
   Environment,
   Event,
   PredicateFailure,
   STS,
   Signal,
   State,
+  TRC (TRC),
+  TransitionRule,
+  judgmentContext,
   transitionRules,
+  (?!),
  )
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
 import GHC.Generics (Generic)
 import NoThunks.Class (NoThunks (..))
 
 data ConwayVDelPredFailure era
-  deriving (Show, Eq, Generic, NoThunks, NFData)
+  = ConwayDRepAlreadyRegisteredVDEL !(Credential 'Voting (EraCrypto era))
+  | ConwayCommitteeAlreadyRegisteredVDEL !(KeyHash 'CommitteeColdKey (EraCrypto era))
+  deriving (Show, Eq, Generic)
 
-data VDelEnv era = VDelEnv
+instance NoThunks (ConwayVDelPredFailure era)
+
+instance NFData (ConwayVDelPredFailure era)
 
 newtype ConwayVDelEvent era = VDelEvent (Event (EraRule "VDEL" era))
 
 instance
-  ( Era era
+  ( EraPParams era
   , State (EraRule "VDEL" era) ~ VState era
-  , Signal (EraRule "VDEL" era) ~ ConwayCommitteeCert (EraCrypto era)
-  , Environment (EraRule "VDEL" era) ~ VDelEnv era
+  , Signal (EraRule "VDEL" era) ~ ConwayVDelCert (EraCrypto era)
+  , Environment (EraRule "VDEL" era) ~ PParams era
   , EraRule "VDEL" era ~ ConwayVDEL era
   , Eq (PredicateFailure (EraRule "VDEL" era))
   , Show (PredicateFailure (EraRule "VDEL" era))
   ) =>
   STS (ConwayVDEL era)
   where
   type State (ConwayVDEL era) = VState era
-  type Signal (ConwayVDEL era) = ConwayCommitteeCert (EraCrypto era)
-  type Environment (ConwayVDEL era) = VDelEnv era
+  type Signal (ConwayVDEL era) = ConwayVDelCert (EraCrypto era)
+  type Environment (ConwayVDEL era) = PParams era
   type BaseM (ConwayVDEL era) = ShelleyBase
   type PredicateFailure (ConwayVDEL era) = ConwayVDelPredFailure era
   type Event (ConwayVDEL era) = ConwayVDelEvent era
 
-  transitionRules = undefined -- TODO
+  transitionRules = [conwayVDelTransition @era]
+
+conwayVDelTransition :: TransitionRule (ConwayVDEL era)
+conwayVDelTransition = do
+  TRC
+    ( _pp
+      , vState@VState {vsDReps, vsCCHotKeys}
+      , c
+      ) <-
+    judgmentContext
+  case c of
+    ConwayVDelRegDRep cred -> do
+      Set.notMember cred vsDReps ?! ConwayDRepAlreadyRegisteredVDEL cred
+      -- NOTE: @Soupstraw: something about DRep Deposits seems to be missing here.
+      pure $
+        vState
+          { vsDReps = Set.insert cred vsDReps
+          }
+    ConwayVDelUnRegDRep cred ->
+      pure $
+        vState
+          { vsDReps = Set.delete cred vsDReps
+          }
+    ConwayVDelRegCommittee coldK hotK -> do
+      -- NOTE: @Soupstraw: from the spec: "Should we check if KeyHash actually belongs to the CC?"
+      Map.notMember coldK vsCCHotKeys ?! ConwayCommitteeAlreadyRegisteredVDEL coldK
```

Comment:
We might want to allow re-registering an already exisiting key, the spec is not very clear about this. I'll look into it.

Reply:
yeah, it seems like we'll want to allow re-registering without having to unregister first, so this check can be removed.